File: ResumeApp/Controllers/ReferenceController.cs
using Microsoft.AspNetCore.Mvc;
using Google.Cloud.Firestore;
using ResumeApp.Models.Entities;

namespace ResumeApp.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class ReferenceController : ControllerBase
    {
        private readonly FirestoreDb _firestoreDb;
        private readonly ILogger<ReferenceController> _logger;

        public ReferenceController(FirestoreDb firestoreDb, ILogger<ReferenceController> logger)
        {
            _firestoreDb = firestoreDb;
            _logger = logger;
        }

        // GET: api/reference - Get approved references only
        [HttpGet]
        public async Task<IActionResult> GetApprovedReferences()
        {
            try
            {
                CollectionReference collection = _firestoreDb.Collection("references");
                Query query = collection.WhereEqualTo("approved", true);
                QuerySnapshot snapshot = await query.GetSnapshotAsync();

                var references = snapshot.Documents.Select(doc =>
                {
                    var reference = doc.ConvertTo<Reference>();
                    return new
                    {
                        id = doc.Id,
                        name = reference.Name,
                        title = reference.Title,
                        company = reference.Company,
                        relationship = reference.Relationship,
                        message = reference.Message,
                        rating = reference.Rating,
                        createdAt = reference.CreatedAt.ToDateTime()
                    };
                })
                .OrderBy(r => GetRelationshipPriority(r.relationship))
                .ThenByDescending(r => r.createdAt)
                .ToList();

                return Ok(references);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving approved references");
                return StatusCode(500, new
                {
                    success = false,
                    message = "An error occurred while retrieving references."
                });
            }
        }

        private int GetRelationshipPriority(string? relationship)
        {
            return relationship?.ToLower() switch
            {
                "manager" => 1,
                "direct-report" => 2,
                "colleague" => 3,
                "client" => 4,
                "mentor" => 5,
                "other" => 6,
                _ => 99
            };
        }

                [HttpGet("all")]
        public async Task<IActionResult> GetAllReferences()
        {
            try
            {
                CollectionReference collection = _firestoreDb.Collection("references");
                QuerySnapshot snapshot = await collection.GetSnapshotAsync();

                var references = snapshot.Documents.Select(doc =>
                {
                    var reference = doc.ConvertTo<Reference>();
                    return new
                    {
                        id = doc.Id,
                        name = reference.Name,
                        title = reference.Title,
                        company = reference.Company,
                        email = reference.Email,
                        relationship = reference.Relationship,
                        message = reference.Message,
                        rating = reference.Rating,
                        approved = reference.Approved,
                        createdAt = reference.CreatedAt.ToDateTime()
                    };
                })
                .OrderByDescending(r => r.createdAt)
                .ToList();

                return Ok(references);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error retrieving all references");
                return StatusCode(500, new
                {
                    success = false,
                    message = "An error occurred while retrieving references."
                });
            }
        }

                [HttpPost]
        public async Task<IActionResult> SubmitReference([FromBody] Reference reference)
        {
            try
            {
                if (reference == null)
                {
                    return BadRequest("Reference data is required");
                }

                // Validate required fields
                if (string.IsNullOrWhiteSpace(reference.Name) ||
                    string.IsNullOrWhiteSpace(reference.Email) ||
                    string.IsNullOrWhiteSpace(reference.Relationship) ||
                    string.IsNullOrWhiteSpace(reference.Message))
                {
                    return BadRequest("Name, Email, Relationship, and Message are required");
                }

                // Set metadata
                reference.CreatedAt = Timestamp.GetCurrentTimestamp();
                reference.Approved = false; // All new references need approval

                // Save to Firestore
                CollectionReference collection = _firestoreDb.Collection("references");
                DocumentReference docRef = await collection.AddAsync(reference);

                _logger.LogInformation($"Reference submitted by {reference.Name} with ID: {docRef.Id}");

                return Ok(new
                {
                    success = true,
                    message = "Thank you for your reference! It will be reviewed and published soon.",
                    referenceId = docRef.Id
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error submitting reference");
                return StatusCode(500, new
                {
                    success = false,
                    message = "An error occurred while submitting your reference. Please try again."
                });
            }
        }

        // PATCH: api/reference/{id}/approve - Approve a reference
        [HttpPatch("{id}/approve")]
        public async Task<IActionResult> ApproveReference(string id)
        {
            try
            {
                DocumentReference docRef = _firestoreDb.Collection("references").Document(id);
                DocumentSnapshot snapshot = await docRef.GetSnapshotAsync();

                if (!snapshot.Exists)
                {
                    return NotFound(new { success = false, message = "Reference not found" });
                }

                await docRef.UpdateAsync("approved", true);

                _logger.LogInformation($"Reference {id} approved");

                return Ok(new
                {
                    success = true,
                    message = "Reference approved successfully"
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Error approving reference {id}");
                return StatusCode(500, new
                {
                    success = false,
                    message = "An error occurred while approving the reference."
                });
            }
        }

                [HttpPatch("{id}/reject")]
        public async Task<IActionResult> RejectReference(string id)
        {
            try
            {
                DocumentReference docRef = _firestoreDb.Collection("references").Document(id);
                DocumentSnapshot snapshot = await docRef.GetSnapshotAsync();

                if (!snapshot.Exists)
                {
                    return NotFound(new { success = false, message = "Reference not found" });
                }

                await docRef.UpdateAsync("approved", false);

                _logger.LogInformation($"Reference {id} rejected");

                return Ok(new
                {
                    success = true,
                    message = "Reference rejected successfully"
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Error rejecting reference {id}");
                return StatusCode(500, new
                {
                    success = false,
                    message = "An error occurred while rejecting the reference."
                });
            }
        }

                [HttpDelete("{id}")]
        public async Task<IActionResult> DeleteReference(string id)
        {
            try
            {
                DocumentReference docRef = _firestoreDb.Collection("references").Document(id);
                DocumentSnapshot snapshot = await docRef.GetSnapshotAsync();

                if (!snapshot.Exists)
                {
                    return NotFound(new { success = false, message = "Reference not found" });
                }

                await docRef.DeleteAsync();

                _logger.LogInformation($"Reference {id} deleted");

                return Ok(new
                {
                    success = true,
                    message = "Reference deleted successfully"
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Error deleting reference {id}");
                return StatusCode(500, new
                {
                    success = false,
                    message = "An error occurred while deleting the reference."
                });
            }
        }
    }
}

File: JavaSpring/src/main/java/com/resumeapp/controller/ReferenceController.java
package com.resumeapp.controller;

import com.google.api.core.ApiFuture;
import com.google.cloud.Timestamp;
import com.google.cloud.firestore.*;
import com.resumeapp.model.Reference;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/references")
@CrossOrigin(origins = "*")
public class ReferenceController {

    @Autowired
    private Firestore firestore;

    @GetMapping("/approved")
    public List<Reference> getApprovedReferences() throws ExecutionException, InterruptedException {
        ApiFuture<QuerySnapshot> future = firestore.collection("references")
                .whereEqualTo("approved", true)
                .get();

        List<QueryDocumentSnapshot> documents = future.get().getDocuments();
        List<Reference> references = documents.stream()
                .map(doc -> doc.toObject(Reference.class))
                .collect(Collectors.toList());

        references.sort((a, b) -> {
            int priorityA = getRelationshipPriority(a.getRelationship());
            int priorityB = getRelationshipPriority(b.getRelationship());

            if (priorityA != priorityB) {
                return Integer.compare(priorityA, priorityB);
            }

            return b.getCreatedAt().compareTo(a.getCreatedAt());
        });

        return references;
    }

    @GetMapping
    public List<Reference> getAllReferences() throws ExecutionException, InterruptedException {
        ApiFuture<QuerySnapshot> future = firestore.collection("references").get();

        List<QueryDocumentSnapshot> documents = future.get().getDocuments();
        return documents.stream()
                .map(doc -> doc.toObject(Reference.class))
                .sorted((a, b) -> b.getCreatedAt().compareTo(a.getCreatedAt()))
                .collect(Collectors.toList());
    }

    @PostMapping
    public ResponseEntity<Map<String, String>> submitReference(@RequestBody Reference reference)
            throws ExecutionException, InterruptedException {
        reference.setApproved(false);
        reference.setCreatedAt(Timestamp.now());

        ApiFuture<DocumentReference> future = firestore.collection("references").add(reference);
        DocumentReference docRef = future.get();

        Map<String, String> response = new HashMap<>();
        response.put("id", docRef.getId());
        response.put("message", "Reference submitted successfully");

        return ResponseEntity.ok(response);
    }

    @PatchMapping("/{id}/approve")
    public ResponseEntity<Map<String, String>> approveReference(@PathVariable String id)
            throws ExecutionException, InterruptedException {
        DocumentReference docRef = firestore.collection("references").document(id);

        Map<String, Object> updates = new HashMap<>();
        updates.put("approved", true);

        ApiFuture<WriteResult> future = docRef.update(updates);
        future.get();

        Map<String, String> response = new HashMap<>();
        response.put("message", "Reference approved successfully");

        return ResponseEntity.ok(response);
    }

    @PatchMapping("/{id}/reject")
    public ResponseEntity<Map<String, String>> rejectReference(@PathVariable String id)
            throws ExecutionException, InterruptedException {
        DocumentReference docRef = firestore.collection("references").document(id);

        Map<String, Object> updates = new HashMap<>();
        updates.put("approved", false);

        ApiFuture<WriteResult> future = docRef.update(updates);
        future.get();

        Map<String, String> response = new HashMap<>();
        response.put("message", "Reference rejected successfully");

        return ResponseEntity.ok(response);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, String>> deleteReference(@PathVariable String id)
            throws ExecutionException, InterruptedException {
        ApiFuture<WriteResult> future = firestore.collection("references").document(id).delete();
        future.get();

        Map<String, String> response = new HashMap<>();
        response.put("message", "Reference deleted successfully");

        return ResponseEntity.ok(response);
    }

    private int getRelationshipPriority(String relationship) {
        if (relationship == null)
            return 999;

        switch (relationship.toLowerCase()) {
            case "manager":
                return 1;
            case "direct-report":
                return 2;
            case "colleague":
                return 3;
            case "client":
                return 4;
            case "mentor":
                return 5;
            default:
                return 6;
        }
    }
}
